File: elixir/1524_num_of_subarrays.exs
# https://leetcode.com/problems/number-of-sub-arrays-with-odd-sum/
# 25/02/2025

defmodule Solution do
  @mod 1_000_000_000 + 7

  @spec num_of_subarrays(arr :: [integer]) :: integer
  def num_of_subarrays(arr) do
    arr
    |> generate_prefix_sum([])
    |> count_odd_sums(0, 0, 0)
  end

  defp generate_prefix_sum([], acc), do: Enum.reverse(acc)
  defp generate_prefix_sum([num | rest], []), do: generate_prefix_sum(rest, [num])

  defp generate_prefix_sum([num | rest], acc),
    do: generate_prefix_sum(rest, [num + hd(acc) | acc])

  defp count_odd_sums([], result, _even, _odd), do: result

  defp count_odd_sums([curr_ps | rest], result, even, odd) do
    case(rem(curr_ps, 2)) do
      0 -> count_odd_sums(rest, rem(result + odd, @mod), even + 1, odd)
      1 -> count_odd_sums(rest, rem(result + even + 1, @mod), even, odd + 1)
    end
  end
end

# defmodule Solution do
#   @mod 1_000_000_000 + 7

#   @spec num_of_subarrays(arr :: [integer]) :: integer
#   def num_of_subarrays(arr) do
#     arr
#     |> Enum.scan(0, &(&1 + &2))
#     |> Enum.reduce({0, 0, 0}, fn curr_ps, {result, even, odd} ->
#       case(rem(curr_ps, 2)) do
#         0 -> {rem(result + odd, @mod), even + 1, odd}
#         1 -> {rem(result + even + 1, @mod), even, odd + 1}
#       end
#     end)
#     |> elem(0)
#   end
# end

defmodule Main do
  def main do
    test_cases = [
      {[1, 3, 5], 4},
      {[2, 4, 6], 0},
      {[1, 2, 3, 4, 5, 6, 7], 16}
    ]

    test_cases
    |> Enum.each(fn {nums, expected} ->
      result = Solution.num_of_subarrays(nums)
      IO.puts("Input: #{inspect(nums)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/1524_num_of_subarrays.rs
// https://leetcode.com/problems/number-of-sub-arrays-with-odd-sum/
// 25/02/2025

const MOD: i32 = 1_000_000_000 + 7;

impl Solution {
    pub fn num_of_subarrays(arr: Vec<i32>) -> i32 {
        arr.iter()
            .scan(0, |acc, num| {
                *acc += num;
                Some(*acc)
            })
            .fold((0, 0, 0), |(result, ces, cos), prefix_sum| {
                match prefix_sum % 2 {
                    0 => ((result + cos) % MOD, ces + 1, cos),
                    1 => ((result + ces + 1) % MOD, ces, cos + 1),
                    _ => unreachable!(),
                }
            })
            .0
    }
}

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case([1, 3, 5].to_vec(), 4)]
    #[case([2, 4, 6].to_vec(), 0)]
    #[case([1, 2, 3, 4, 5, 6, 7].to_vec(), 16)]
    fn test(#[case] nums: Vec<i32>, #[case] expected: i32) {
        let result = Solution::num_of_subarrays(nums);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        ([1, 3, 5].to_vec(), 4),
        ([2, 4, 6].to_vec(), 0),
        ([1, 2, 3, 4, 5, 6, 7].to_vec(), 16),
    ];

    for (nums, expected) in test_cases {
        let result = Solution::num_of_subarrays(nums.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            nums, expected, result
        );
        assert_eq!(result, expected);
    }
}
